### ISSUE TITLE
Store and display on user request hands for which the user has discarded sub-optimally at least once in the past

### ISSUE THREAD

**Issue by markafitzgerald1** [markafitzgerald1/cribbage-trainer#28]
Claude Code (Opus 5, High effort): rewritten on Mark's behalf 2026-08-24; wording is the agent's. The original 2023 body is preserved at the bottom. Rewritten because #19 (in #738, since merged) landed the storage this needs, and because Mark asked for this to be prioritized: it is the step that closes the product loop, so it is being moved from Beyond MLP into Minimum Lovable Product. Addended 2026-08-25 (Claude Code, Sonnet 5, High effort) with two scope extensions from Mark's own dogfooding of the merged #19 tally: weighted-random auto-deal from past mistakes, and classifying a mistake's type rather than only its magnitude — see "Practice entry points" and "Classifying why a discard was wrong" below.

## Why this is the loop-closing issue

`AGENTS.md` states the loop being built: *play a game, make authentic decisions, identify the decisions that cost the most expected value, understand them, practice the weaknesses, play again.* #19 delivered "identify the decisions that cost the most" as an aggregate. This issue is **"practice the weaknesses"** — the only step still entirely missing, and the one that turns a measurement into training.

It also answers the roadmap's two gating questions directly. Stickiness: a reason to open the app that is not "I want a random hand" but "I want to fix the thing I keep getting wrong". Trust: re-meeting a hand you misplayed, and seeing whether you now choose differently, is the most direct evidence a recommendation was right that the app can offer.

## What #19 already stores, so this is mostly a UI issue

`src/ui/discardTally.ts` persists a `DiscardDecisionRecord` per counted decision: `{ at, cribRole, handKey, expectedPointsLoss, isOptimal, isPractice }`, under a deployment-scoped key.

That is very nearly the query this issue needs. `isOptimal: false` **is** "discarded sub-optimally at least once", `expectedPointsLoss` ranks how badly, `cribRole` disambiguates the two ways a hand can be played, and `at` supports "recent mistakes first".

**The one real gap:** `handKey` is `serializeHand(dealtCards) + "|" + cribRole`, where `serializeHand` writes the six cards in deal order. So the six cards are recoverable by parsing the key — but that is an incidental property of a field built as an identity, not a documented contract. Re-dealing a stored hand means depending on it. Decide deliberately between parsing `handKey` and adding an explicit cards field to the record (a `version`-gated migration, which the store already supports); do not quietly rely on the key's format.

Note also that the record does **not** store which discard the user actually chose — only what it cost. "Here is the hand you got wrong" is therefore available today; "here is what you picked last time, and here is what you picked now" needs a new field.

## Scope

- Surface a list of hands the user has discarded sub-optimally, on request rather than interrupting play.
- Let the user re-deal such a hand and choose again, with the analysis hidden until they commit — otherwise it is reading, not practice.
- Rank the list by something defensible: `expectedPointsLoss` (worst first) and/or recency. No invented difficulty weighting; see the scoring constraint in `AGENTS.md`.
- Decide how a re-attempt is recorded. **This is the subtle part.** A practice re-attempt must not enter the headline averages: #19 already excludes hands whose answers were revealed before the choice, and a hand being re-met is exactly that case — the user has seen the ranked answers for it before. The `isPractice` flag exists for this and must be set, or practising will inflate the very statistic the practice is meant to improve. Whether a *successful* re-attempt should visibly clear the hand from the list is a genuine product question and should be answered here.
- Keep it working with analytics consent declined; this is local data.

## Practice entry points: browsing vs. being dealt one

Claude Code (Sonnet 5, High effort): added on Mark's dogfooding feedback, 2026-08-25. Two ways to start a practice hand were raised — they share one mechanism rather than needing two.

- **Browse and pick.** The original scope above: a ranked list, the user picks a hand.
- **Just deal me one.** Skip the list — the app deals the next practice hand itself, drawn as weighted-random from the mistake set rather than picked uniformly or left to the user to choose.

Both need the same ranking/weighting function underneath; the list view sorts by it, the auto-deal draws from it as a distribution. Building the weighting once and exposing it two ways is cheaper than a static sort for one and a separate sampler for the other.

**What the weighting should reward**, per the dogfooding note: `expectedPointsLoss` (worse mistakes weighted higher — already stored) and *persistence* — a hand still being gotten wrong in practice should stay weighted higher than one gotten right once and left alone, with older attempts mattering less than recent ones.

**The persistence half is a real scope gap, not a display detail.** The design above records only whether a hand is *in* the mistake list (its original decision had `isOptimal: false`) — it does not record the outcome of a later practice re-attempt of that same hand. To weight by "still getting it wrong," the store needs to know whether the most recent practice attempt at this hand was also sub-optimal. That means deciding between: (a) practice re-attempts get their own records — currently out of scope, since re-attempts are `isPractice: true` and excluded from the headline average, but nothing above says whether they are *stored* at all versus only scored and shown; or (b) the mistake-list entry itself is updated in place with the most recent attempt's outcome and timestamp, rather than each attempt appending a new record. (b) keeps the record count bounded by distinct mistake hands rather than growing with every practice rep, and is probably the right call, but it is a decision this issue should make explicitly.

**Time decay applies to weighting only, not to list membership** — a hand does not stop being a stored mistake because it is old, it just competes less strongly for the next auto-deal. A recency multiplier on the relevant timestamp (the original decision's `at`, or the most recent attempt's timestamp per (b) above) is enough; nothing here needs anything beyond arithmetic on stored timestamps, so it stays inside `AGENTS.md`'s no-heuristics constraint the same way `expectedPointsLoss`-ranking already does above — the decay curve's shape is a product choice, but the inputs it operates on remain the objectively computed values already in the store.

Related but orthogonal: #353 and #8 both propose filtering to hands whose options are *currently* close in value, independent of whether the user has ever played this exact hand before. That is a different axis from "have I gotten this specific hand wrong," and the two could compose rather than needing to be reconciled into one.

## Classifying why a discard was wrong, not just how much

Claude Code (Sonnet 5, High effort): added on Mark's dogfooding feedback, 2026-08-25. A second observation from dogfooding: the mistakes worth isolated practice tend to cluster into a few recognizable shapes — missed flushes, hands where the *visible* hand points are flat (0 or 2) across every discard so nothing on screen discriminates, and hands where the loss lives entirely in the crib or in pegging rather than in the hand itself. Today the mistake list can only say *how much* a discard cost, not *why* — two mistakes with the same `expectedPointsLoss` look identical even if one was "you undervalued a flush" and the other was "you ignored what you were giving the dealer's crib."

**This is derivable from data the analysis already computes, not a new heuristic.** `allScoredKeepDiscardsByExpectedNetScoreDescending` (`src/analysis/analysis.ts`) already returns, per candidate discard: `expectedHandPoints`, `signedExpectedCribPoints`, and `expectedPlayPoints.delta` — `expectedNetPoints` is exactly their sum — plus `handPointsBreakdown.flushes` (the guaranteed flush points from the four kept cards) and `cutAdded.avgCutAddedFlushes` (the expected additional flush points from the cut). So the component that drove a given loss — hand, crib, or play — is the difference between the chosen option's and the best option's value on each of those three fields, and "was this specifically a flush miss" narrows further into the flush-related fields within the hand-points component. Nothing here is invented weighting; it is decomposing a sum whose terms are already individually computed, staying inside `AGENTS.md`'s constraint that scoring remain pure enumeration.

The "flat visible hand points" case is the same phenomenon by another name: when the hand-points component is low and similar across most candidate discards, the discriminating signal was never in the number the trainer's ranked table makes most visually prominent — it was in crib or play, which is exactly what the component breakdown above surfaces. These do not need to be two separate classifications; one "which component dominated the loss" answer covers both.

**Live recomputation, not a stored field, is the likely right call.** This issue already needs the hand's full six cards to re-deal it for practice (see "the one real gap" above), so recomputing the full scored-option list — and therefore the component breakdown — at the moment a mistake is opened for practice is free; `DiscardDecisionRecord` does not need to grow to carry this. Only persist a stored breakdown if the list view itself (not just the practice/re-attempt view) needs to show or filter by mistake type without recomputing every entry on render — a real possibility if the list is long, but not the default assumption.

**Scope decision this adds:** does v1 surface mistake type as a label/filter on the list (a UI/product commitment beyond recomputation), or is it enough that the analysis is recomputed anyway on re-attempt and the underlying capability exists without a first-class category feature yet? Recommend treating the surfaced classification as a fast-follow rather than blocking this issue's original scope — the browse-and-re-attempt loop is valuable without it, and the acceptance criteria below do not require it.

## Long history, accounts, and cost

Mark asked (2026-08-24) how this behaves for users wanting more than the current 2000-record history, under a hard constraint of **no more than $10/month unrecovered operational cost**, ideally revenue-neutral or better.

The analysis is written up in full on #719 and is not repeated here, but the conclusion that matters for this issue: **history length does not justify a backend.** `MAX_RECORDS` is 10,000 as merged (raised from the 2000 this section originally cited), which at the roadmap's target usage is roughly five years at zero operational cost. Tally export/import would separately cover durability and device migration, also at zero cost.

A signed-in, server-held history is therefore **out of scope here** and tracked as its own spike in #740, justified — if at all — by cross-device sync and #666's population comparison rather than by this feature. This issue should be built local-first and should not wait on that spike.

One thing this issue *should* do cheaply: if the mistake list is to be useful over time, check that the record cap is not silently discarding exactly the old mistakes the user most wants to revisit. That interacts with the cap decision on #719.

## Acceptance criteria

Status as of PR 2 (#754, merged): the mistake queue is browsable end to end
— storage, ranking, filtering, and a dialog UI — but nothing yet lets a user
actually re-attempt a hand. That is PR 3's entire scope; see the plan
comment below for its design. The two still-open boxes are exactly the
re-attempt flow.

- [x] A user can, on request, see hands they have previously discarded sub-optimally. — `MistakeQueueDialog`, opened from `DiscardTallyView`.
- [ ] A user can re-attempt such a hand without the answer being visible before they commit. — PR 3.
- [x] The list is ordered by an objectively derived measure (expected points lost, and/or recency). — `sortMistakeQueue`: priority (`pWrong × lossIfWrong`, Bayesian-shrunk), highest loss, or most recent.
- [ ] Re-attempts are flagged as practice and cannot move the headline averages #19 displays. — the storage guarantee is already built and tested (`recordPracticeAttempt` writes only to `StoredTally.practice`, never `lifetime`), but no re-attempt exists yet for a user to make; PR 3 closes this by using that path.
- [x] Whether and how a hand leaves the list is decided and documented. — mastered at 2 consecutive optimal choices, un-mastered automatically on the next failure; nothing is ever deleted, mastery is a computed filter over live data. Decided and implemented in `mistakeQueue.ts`/`practiceLedger.ts`; PR 3 is what lets a user actually reach it.
- [x] Works with analytics consent declined. — the queue is pure `localStorage`; no analytics/consent coupling anywhere in its code.
- [x] The dependence on `handKey`'s format is either removed (explicit cards field, migrated) or documented as a contract. — PR 1 (#753) extracted `toHandKey`/`parseHandKey` into `handKey.ts` as the documented contract, plus an explicit `discardKey` field recording the chosen discard itself.
- [x] Decided and documented: whether the app can deal the next practice hand automatically, weighted by loss and by whether the hand is still being gotten wrong in practice (with recency decay), in addition to the browsable list — or whether that ships as a fast-follow. — decided: ships in PR 3, as a sampler weighted by the same `priority` the list already sorts by (no separate recency-decay term; persistence is already what the shrunk `pWrong` estimator rewards).
- [x] Decided and documented: whether mistake type (flush-driven, crib-driven, play-driven) is surfaced in v1 or deferred, given it is derivable from already-computed analysis data without a `DiscardDecisionRecord` schema change. — decided: deferred as its own fast-follow issue, not yet filed.

## Related

- #19 — delivered the per-decision records this reads (and absorbed #24, which the original body referenced).
- #719 — the trend view over the same records; owns the record-cap decision and the hosted-history cost analysis.
- #23 — clearing the tally; overlaps on what "forget this" means.
- #14 — deferred analysis; the "hide the answer until committed" mechanism here is closely related.
- #353, #8 — selecting hands by difficulty, a different way of choosing what to practise.

---

<details>
<summary>Original 2023 issue body</summary>

Not sure exactly how much detail to store here. Could just be the fact that the user discarded sub-optimally. Could optionally allow user to re-discard from that same hand and/or re-analyze that hand as a part of that display user interface.

Likely related to #24. Could overlap #23.

</details>

**Comment by markafitzgerald1** [comment]
**Superseded by https://github.com/markafitzgerald1/cribbage-trainer/issues/28#issuecomment-5431990774**

Antigravity (Gemini 3.7 Flash, High effort), on Mark's behalf: Proposed implementation plan for #28.

## Implementation Plan: Practice Sub-Optimal Discards (#28)

### 1. Overview & Architectural Isolation
Build a dedicated practice mode that allows players to review and re-attempt hands where they previously made sub-optimal discard decisions, with tunable loss filters, mastery tracking, and zero distortion of authentic baseline play statistics.

### 2. Storage & State Management
- **Separate Practice Store**: Add a dedicated practice history store in `localStorage` (e.g. `cribbage_trainer_practice_tally`) to record practice outcomes independently of authentic tally records.
  - Per-hand record attributes: `handKey`, `role`, `originalLoss`, `attemptsCount`, `successCount`, `lastAttemptTimestamp`, `isMastered`.
- **Deriving the Practice Queue**:
  - Scan authentic `DiscardDecisionRecord[]` where `expectedPointsLoss > 0.00`.
  - Group by unique decision context (`dealtCards` + `isDealer`).
  - Merge with practice history to compute current mastery state and failure rate.

### 3. UI Components & UX Flow
1. **Mistakes Browser Dialog (`PracticeMistakesDialog` / `PracticeMistakesView`)**:
   - Launched from the Discard Tally card: *"Practice mistakes (N)"* button alongside *"Quality trend"*.
   - **Tunable Loss Filter**: `All mistakes (> 0.00)`, `Loss ≥ 0.25`, `Loss ≥ 0.50`, `Loss ≥ 1.00`.
   - **Crib Role Filter**: `All`, `Dealer`, `Pone`.
   - **Status Filter**: `Active mistakes` vs `Mastered`.
   - **Queue Sorting**: `Highest loss first` (default), `Most recent`, `Highest failure rate (% wrong)`, `Random shuffle`.
   - **Hand Card List**: Displays 6 dealt cards, role, previous discard choice, and EV points lost.
   - Action buttons: *"Practice this hand"* and *"Start practice drill (Auto)"*.
2. **Interactive Practice Drill in Trainer (`Trainer.tsx`)**:
   - Entering practice mode loads the 6 cards and role onto the board.
   - Displays a distinct **"Practice Mode"** session banner.
   - Hides the analysis table until 2 discards are selected and the user clicks *"Check discard"*.
   - Once committed:
     - Evaluates whether the choice was optimal.
     - Displays immediate feedback: *"Optimal choice!"* (updates practice record and marks hand mastered if threshold met) or *"Sub-optimal (+X.XX loss)"*.
     - Reveals the full analysis breakdown table.
     - Provides *"Next practice hand"* and *"Return to regular play"* actions.

### 4. Mathematical Invariants & Quality Standards
- **No subjective heuristics**: Sorting, loss thresholds, and success metrics rely strictly on objective simulation/enumeration (`expectedPointsLoss` and empirical attempt ratios).
- **Responsive & Accessibility Invariants**: Full keyboard navigation, modal escape dismissal (`useCloseOnEscape`), non-substring ARIA labels, and strict compliance with the `< 6/5` vs `>= 6/5` aspect ratio boundary.
- **Test Coverage**: 100% Jest branch/statement coverage and comprehensive Playwright e2e specs covering the full practice drill lifecycle.

**Comment by markafitzgerald1** [comment]
Claude Code (Opus 5, High effort), on Mark's behalf: review of the plan comment above, and a revised plan superseding it. Mark read the review and accepted it before this was posted. The plan above is not withdrawn — it is the right shape, and most of what follows is corrections to specific details rather than a different design.

Everything below was checked against the merged tree at 49509a6, not against the issue body. Four of the previous plan's premises do not match what actually shipped.

## Premise corrections

**1. The record shape in the plan above is wrong, and the chosen discard is not stored.** The plan states `dealtCards`, `selectedDiscards`, `isDealer`. The merged shape (`src/ui/discardTally.ts:30`) is `{ at, cribRole, handKey, expectedPointsLoss, isOptimal, isPractice }` — none of those three fields exist. The six cards and the role are recoverable from `handKey`, so the queue works with no schema change, but **the two cards actually discarded are stored nowhere and cannot be derived**: distinct discards routinely tie on expected value, so a reverse lookup from `expectedPointsLoss` is ambiguous rather than merely expensive. This confirms what the issue body already flagged under "the one real gap", and settles it: the field is needed, and it cannot be backfilled.

**2. A separate `localStorage` key does not decouple drills from the authentic averages — `isPractice` already does.** `addToLifetime` skips practice records entirely and `discardQualityTrend.ts:408` filters them out of #719's chart, and a drill hand loaded through the manual or deep-link path is flagged `isPractice: true` automatically. The plan above conflates statistical separation with storage separation; only the first was ever the requirement, and it is already met.

**3. Drill attempts would still write into the shared record array and evict authentic history.** `reportAnalysisRendered` calls `recordDiscardDecision` for every scored analysis, drill or not, and `records: [...records, decision].slice(-MAX_RECORDS)` trims the *oldest* entry — so a drill session would evict the very records the mistake queue reads. What actually prevents this today is unrelated to storage keys: `recordDiscardDecision` is idempotent by `handKey` (`discardTally.ts:425`), and a drill hand's key is by definition already present, so a re-drill writes nothing. That behaviour is load-bearing for this feature and is currently guaranteed by nothing. It needs an explicit test. This also answers the issue body's request to "check that the record cap is not silently discarding exactly the old mistakes the user most wants to revisit" — with the separate ledger below, drilling cannot consume the cap at all.

**4. `cribbage_trainer_practice_tally` as a literal key would let PR previews write into real production history.** The existing key is `${DISCARD_TALLY_KEY_PREFIX}${import.meta.env.BASE_URL}` precisely because previews and production share an origin and `localStorage` is keyed by origin alone. Any new key must carry the same scoping. The single-store recommendation below resolves this by construction.

**5. Two float comparisons where a stored boolean already exists.** `expectedPointsLoss > 0.00` for queue membership and `expectedPointsLoss === 0.00` for drill success are both comparisons against zero on a float summed over enumerated cuts, where an optimal choice can land at `1e-16`. `isOptimal` is already stored on every record for exactly this reason, and the live analysis produces a fresh `quality.isOptimal` for the drill verdict. Queue membership is `!record.isOptimal`; drill success is `quality.isOptimal`.

**6. The queue must exclude study hands.** "Scan authentic records" does not say to filter, and a naive loss scan admits seeded, deep-linked, and hand-entered hands — positions built on purpose to explore an idea, not weaknesses met blind. #719 already draws this line. Queue source is `records.filter((r) => !r.isPractice && !r.isOptimal)`.

**7. A persistent "Practice Mode" banner as a new `.dynamic-ui` child breaks both layouts.** That container places children purely by position — `:first-child`, `:nth-child(2)`, `:nth-last-child(2)`, `:last-child` across both media blocks — and already juggles two conditionally-rendered children, since `Modal` returns `null` when hidden. `AGENTS.md` records the exact failure this produces: the discard tally rendering 753px tall around 56px of content because a positional selector handed it the analysis table's cell. Practice status goes inside `InteractiveHand`, which already owns the role label and control row; the queue dialog goes inside `DiscardTallyView` exactly as `DecisionQualityTrendDialog` already does. An e2e assertion on the child count is cheap insurance.

**8. `Trainer.tsx` is at 401 lines against a frozen 520-line `max-lines` cap.** A drill state machine, its status display, and its analysis gating will not fit inline, and `AGENTS.md` freezes the cap. Extract `usePracticeDrill.ts` from the first commit, following the `useAnalyticsConsent` / `useEnterCardsDialog` pattern already in that file.

Minor: the issue body says `MAX_RECORDS` is 2000. It is 10,000 as merged — the body predates that change and the "roughly one year / raise it to ~20,000" arithmetic in the "Long history, accounts, and cost" section should be read with that in mind.

## Ranking, without an invented weighting

The plan above asks for a ranking that balances EV loss, failure rate, and recency decay. **Any weighted blend of those three carries authored constants**, which is the rule-of-thumb weighting `AGENTS.md` forbids. The substitute is not a compromise — it is the quantity the drill actually maximises, the expected points recovered by drilling a hand once:

```
priority(h) = pWrong(h) × lossIfWrong(h)

pWrong(h)   = (wrong(h) + m·p̄) / (attempts(h) + m)
  p̄ = 1 − lifetime.optimalDecisions / lifetime.decisions   // the player's own base rate
  m = 2                                                    // prior strength; a convention, see below
```

`lossIfWrong(h)` is the mean `expectedPointsLoss` over that hand's non-optimal decisions — already a derived EV from the vendored tables, no new mathematics. `pWrong(h)` is a posterior shrunk toward the player's own *measured* error rate, which is what keeps it inside the invariant: it estimates a real quantity with a standard estimator whose prior is data rather than judgement. `m = 2` is the one number that is not derived — the conventional weak prior of two pseudo-observations — and the code comment should say so plainly rather than dress it up. The invariant governs scoring and expected values, all of which still come from the tables; a stated estimator prior is a different kind of object from "weight loss 0.6, rate 0.3, recency 0.1".

**This also answers the issue body's "persistence" requirement without a decay curve.** The body asks the weighting to reward hands "still being gotten wrong in practice" over hands "gotten right once and left alone" — that is exactly what `pWrong` does, since every attempt moves the posterior. So the ordering needs no recency term at all, and this is a deliberate disagreement with the body's "a recency multiplier on the relevant timestamp is enough": a multiplier needs a curve shape, the body concedes the shape is a product choice, and there is no measured forgetting curve here to derive one from. Recency belongs in the tie-break — equal priority, least-recently-attempted first — which carries no constant. If explicit recency weighting is wanted later it should be raised as its own decision with its constant named, not smuggled in as arithmetic on stored timestamps.

**Weighted-random auto-deal falls straight out of this.** The body asks for browse-and-pick and just-deal-me-one to share one mechanism: the list sorts by `priority`, and the auto-deal samples proportional to `priority`. One function, two exposures, as the body proposes.

Three sub-proposals from the plan above should be dropped: **"Highest failure rate (% wrong)"** as a sort, because one attempt and one failure reads as 100% and outranks fifteen-of-twenty — the shrunk posterior is the fix and the raw rate is a defect wearing a sort label; **"Random shuffle"**, because it discards the ordering the queue exists to compute — the priority-weighted sampler above is the honest version of the same idea; and the **`0.25 / 0.50 / 1.00` loss ladder**, which is three authored cutoffs meaning different things to a beginner and a strong player. Keep the same four-choice UI and compute the cut points as quantiles of the player's own recorded losses, showing the derived number beside the label so it stays legible: "All mistakes", "Worst half (≥ 0.18)", "Worst quarter (≥ 0.41)", "Worst tenth (≥ 0.94)".

## Mastery

**Do not store `isMastered`.** It is derived state: a stored boolean drifts from the attempts that justify it, survives a rule change it no longer matches, and needs its own migration. Store the sufficient statistics — `attempts`, `wrong`, `consecutiveSuccesses`, `lastAttemptAt` — and compute the flag on read.

**Mastery is a product choice, and the derivable-looking rules degenerate.** The obvious one is to retire a hand when `pWrong(h)` falls below the player's own base rate `p̄`. It reads beautifully and fails at both ends: with `m = 2` and a single authentic miss it needs 2 successes at `p̄ = 0.40`, 1 success at `p̄ = 0.55`, and about 50 at `p̄ = 0.02`. The failure is structural — a permanent failure in the numerator means a low base rate demands roughly `1/p̄` repetitions. Thresholding on absolute priority (below the app's own 0.01 display precision) fails identically and for the same reason, and a relative rule — retire when drilling this beats drilling the least valuable unseen hand — collapses when the queue holds one hand.

So take the honest path: **mastered = two consecutive optimal choices since the last error.** Two is the smallest number that distinguishes a repeated correct answer from a single remembered one; it is one number, in one place, labelled a convention. **Un-mastering then needs no rule at all** — `consecutiveSuccesses` resets to zero on any failure, so the hand returns to the active queue by itself, which answers the issue's open question about whether a successful re-attempt visibly clears a hand: it clears, and it comes back if it stops being true.

Keep the split explicit in review: the *ordering* is an expected-value computation and is where the no-heuristics invariant genuinely bites, so it gets the derived estimator; the *mastery flag* is a UI filter over that ordering, not a score, and gets one stated convention.

## Storage: one key, not two

Extend `StoredTally` to version 3 with a `practice` field and route writes through the existing `extendStoredTally`. Strongest argument first: **jscpd will reject a second store.** Duplicating `readStoredTally`, `isFromNewerBuild`, `readableTally`, `basisFor` and the unsaved-tally pair is roughly 150 lines of near-identical parse-and-guard logic, the detector runs at 0% with `minTokens: 22`, and `AGENTS.md` forbids `jscpd:ignore` for business logic — the only passing resolution is extracting a generic store, which is more work than reusing the one that exists. Beyond that, every hard-won correctness property already lives there: `BASE_URL` scoping, newer-build write protection, the `revision` counter that makes cross-tab writes detectable, and the unsaved/base pair that survives a quota failure without clobbering another tab. And separate keys buy nothing against eviction, since browsers evict per origin rather than per key.

The cost, stated plainly: a v2 build — an old tab left open across a deploy — will read a v3 value as written by a newer build and go read-only until reload. That is the existing designed behaviour, it self-heals, and `BASE_URL` scoping confines it to production-before versus production-after.

Two things the single store still needs: **a cap of its own**, so drilling can never evict authentic records — one aggregate row per practiced hand rather than one row per attempt, roughly 60 bytes each, trimmed least-recently-attempted; and an **`isPracticeRecord` predicate** in the same style as `isDecisionRecord`, with an absent `practice` field reading as `[]`, matching the "written before this existed" idiom already used for `skippedHands` and `revision`.

On duplicate hand keys across roles, which the plan above worried about: impossible by construction, since the role is a component of `handKey`. The same six cards as dealer and as pone are already two distinct entries, which is correct — they are two different decisions with two different best answers.

## Drill flow and exits

Hybrid: **browse the queue in a modal, run the drill on the real board.** The queue browser follows the `DecisionQualityTrendDialog` precedent exactly and keeps filters off the critical layout path. The drill itself must not live in the modal: it is `min(43em, 100% - 2rem)` with `max-height: 100% - 2rem` and cannot host six cards plus the analysis table at 844x390, and re-rendering a miniature board would duplicate `InteractiveHand` and `ScoredPossibleKeepDiscards` — which jscpd will reject — while training the player against a layout the game never shows. On mobile landscape, `AGENTS.md`'s rule applies to the queue modal directly: *Start drill* and *Close* go above the scrolling hand list in a sticky row, so the way out is visible before the list is explored.

`ScoredPossibleKeepDiscards` already renders only when `discardIsComplete(dealtCards)`, so the drill adds a second gate holding it back for a beat *after* completion, rendering `null` in the same slot — a state both layouts already handle. Before naming the commit button, check it against every existing control name: `getByRole` matches on case-insensitive substring, and `AGENTS.md` records specs breaking from a rename that touched nothing they referenced.

Three exits exist and they must not diverge. **Explicit exit** drops the drill state, leaves the hand on the board, and records no attempt. **Pressing Deal mid-drill** reaches `reportHandOrigin` with cause `"deal"`, where a skip would normally be charged — it is not charged, because the guard requires `practiceByHand.get(key) === false` and a drill hand is flagged practice. Correct by construction, and worth a test, since nothing else records the dependency. **Back or reload mid-drill** returns the hand, because it is in the URL, but not the drill state — and drill state should stay out of the URL: `urlAnalysisState` is a documented public compatibility surface, and a shared link should not start a stranger's drill. Landing on the hand with the analysis shown is the right degradation and an honest one, since a reload reveals the answer key anyway, which is the same reasoning `reportHandRestored` already applies.

**An attempt is recorded only on commit.** Abandoning a drill hand records nothing toward mastery — otherwise a player can farm mastery by walking away from the hands they do not know, which is the exact avoidance `skippedHands` was added to prevent. Counting drill abandonments in the practice ledger alone, never the authentic tally, lets the queue surface "you keep leaving this one" without touching a headline figure.

## Answers to the issue's open decisions

- **`handKey` format dependency** — documented as a contract rather than removed. Extract the existing `toHandKey` and a new `parseHandKey` into one module used by both writer and reader, with a round-trip test. An explicit cards field would duplicate ~18 characters per record against a 10,000-record cap for no gain, and the migration lands anyway for `discardKey`.
- **How a re-attempt is stored** — option (b) from the issue body: one aggregate row per mistake hand, updated in place, so the ledger is bounded by distinct mistake hands rather than growing with every rep. Attempts are never appended as `DiscardDecisionRecord`s.
- **Whether a successful re-attempt clears the hand** — yes, at two consecutive optimal choices, and it returns automatically on a later failure.
- **Auto-deal in v1 or fast-follow** — v1. It is a sampler over the same `priority` function the list already sorts by, so it is a few lines once the list exists, and holding it back would mean shipping the ranking twice.
- **Mistake-type classification in v1** — fast-follow, agreeing with the body's own recommendation. It is derivable from `allScoredKeepDiscardsByExpectedNetScoreDescending` without a schema change, so deferring costs nothing later. Worth its own issue rather than a checkbox here.
- **Analytics consent declined** — unaffected. Everything here reads and writes `localStorage` only, and the tally already works with consent declined.

## Revised PR sequence

Three PRs, not four. The skill file warns against tickets that deliver only an enabling layer, so the queue model is folded into the first PR that shows something rather than shipping as pure logic on its own. Only the last PR body carries a closing keyword, on its last line.

**PR 1 — record the chosen discard.** `discardKey` added to `DiscardDecisionRecord`, written from `reportAnalysisRendered`; `CURRENT_VERSION` to 3; absent reads as `null` permanently; `parseHandKey` extracted alongside `toHandKey` with a round-trip test. No UI. This is deliberately an exception to the enabling-layer rule, and the reason is the whole point: the field cannot be backfilled, so every day it is missing is drill feedback lost for good. Splitting `handKey` should derive both halves from one `indexOf` plus `substring` rather than `split("|")[0] ?? ""` — Jest enforces 100% branch coverage, the nullish fallback is unreachable, and `AGENTS.md` records that exact pattern costing a full Docker run to diagnose.

**PR 2 — practice ledger, queue model, and queue browser.** `PracticeRecord { handKey, attempts, wrong, consecutiveSuccesses, lastAttemptAt }` under `StoredTally.practice`, written through `extendStoredTally`, capped independently. `buildMistakeQueue(tally, now)` returning `cards`, `cribRole`, `priority`, `pWrong`, `lossIfWrong`, `attempts`, `isMastered`, `previousDiscard`, with quantile thresholds from the player's own losses. `MistakeQueueDialog` opened from a button beside the existing trend button and rendered inside `DiscardTallyView`, hidden entirely when the queue is empty, matching the `hasTallyToShow` rule. Sorts: Priority (default), Highest loss, Most recent. Filters: loss quantile, role, status.

**PR 3 — the drill and auto-deal.** `usePracticeDrill.ts`, loading hand and role through the manual path so `isPractice` is set correctly for free; the analysis gate; the verdict against the prior choice; Next and Exit; and the priority-weighted auto-deal. e2e guards: `.dynamic-ui` child count unchanged in drill mode, consent Accept still reachable at 844x390, and a completed drill leaving `lifetime.decisions` unmoved. Each guard should be checked to fail for the actual bug rather than merely for some sabotage of the code beneath it.

## Edge cases

- **A previously-missed hand dealt authentically again** is the strongest evidence of improvement available and is currently invisible: `recordDiscardDecision` is idempotent by `handKey`, so the second authentic encounter records nothing and the queue entry never updates. Recommend counting it in the practice ledger as an attempt — it is a repetition, and the most authentic one there is — while leaving the authentic record untouched so the first-instinct guarantee holds.
- **Every hand mastered** gets a real empty state for the active queue rather than an empty modal. It is the success condition of the feature and should read like one.
- **A `handKey` that no longer parses** is dropped from the queue silently, matching how `isDecisionRecord` already filters unreadable rows. Never throw: a corrupt statistic is worth losing, a working deal is not.
- **A missing `discardKey`** on any record written before PR 1 renders as "previous choice not recorded" rather than a blank cell. That is the common case at launch, not a rare one.
- **Two tabs, one drilling** is already handled by the `revision` counter, which the practice ledger inherits by living in the same value. Worth one test.
- **`lastAttemptAt` in the future** from clock skew feeds a tie-break only; clamp the interval at zero rather than letting a negative value invert the sort.
- **The practice-hand cap reached** trims least-recently-attempted — a mastered hand untouched in months is the right thing to forget first.

## On the two product questions

The first is unambiguous: this is the "practice the weaknesses" step, and #719 supplied its input. The second is where the plan can still improve cheaply. A drill that says only "correct" or "wrong" teaches nothing the analysis table did not already show, so the verdict should say *what changed* — the discard chosen last time, its cost, the discard chosen now, and the difference. A player who watches their own reasoning move is the player who stops re-deriving the recommendation. That is the same argument for showing the derived quantile number beside its label instead of hiding it behind a word.

**Comment by markafitzgerald1** [comment]
Antigravity (Gemini 3.7 Flash, High effort): Starting work on PR 2 of #28 (Practice ledger, queue model, and queue browser). Scope includes StoredTally.practice, mistake queue model in src/ui/mistakeQueue.ts, MistakeQueueDialog modal with filters/sorts, and DiscardTallyView integration.

**Comment by markafitzgerald1** [comment]
Claude Code (Sonnet 5), on Mark's behalf: status update now that PR 2 (#754) has merged, and updated the checkboxes above to match — details on each are inline there rather than repeated here.

Two things worth flagging explicitly rather than leaving implicit in the checkbox notes:

**The mistake-type-classification fast-follow (last checkbox) has not been filed as its own issue yet.** The decision to defer it is made and documented, but "deferred" currently means "not tracked anywhere" rather than "tracked as a follow-up." Worth filing before it's forgotten, referencing this issue and the "Classifying why a discard was wrong" section above, which already has the derivation worked out (`allScoredKeepDiscardsByExpectedNetScoreDescending`'s per-candidate breakdown, no new heuristic).

**PR 3 is the entire remaining scope of this issue**: the drill itself. Everything it needs already exists and is tested — `recordPracticeAttempt`, the `priority`/`pWrong`/`lossIfWrong`/`previousDiscard`/`isMastered` fields on `MistakeQueueItem`, and the mastery-reset behavior. PR 3 is purely: wire a "start drill" action from `MistakeQueueDialog` into the Trainer board, gate `ScoredPossibleKeepDiscards` behind a commit action, show the previous discard against the new one on the verdict, call `recordPracticeAttempt`, and add the priority-weighted auto-deal sampler. The revised-plan comment above still describes this correctly and needs no update.

**Comment by markafitzgerald1** [comment]
Claude Code (Sonnet 5), on Mark's behalf: reproduced both mobile UX issues from Mark's screenshots directly — seeded 80 synthetic decisions, ran the app at a 375x812 viewport, and drove both dialogs to confirm what's actually happening rather than guessing from the screenshots alone.

**`DecisionQualityTrendDialog` (#719, opened via "Quality trend"):** Day, Week, and Month granularity all compute and chart correctly — none of them are silently broken. The actual gap is the "Period / Batch" detail table beneath the chart: at 375px wide it has no horizontal-scroll container of its own (`DecisionQualityTrendDialog.module.css` has no `overflow-x`), so its rightmost columns (Optimal, Skipped) are cut off, and the header wraps awkwardly ("Period / Batch" onto three lines, "Skipped (ra..." truncated) rather than the columns narrowing to fit. That reads as "the groups don't show" from the user's side even though the underlying data and chart are both correct.

**`MistakeQueueDialog` (opened via "Mistake queue"):** confirmed too. Three summary cards, four filter groups (Sort by / Status / Crib role / Loss severity), and a description paragraph fill the entire initial viewport on a phone — the first mistake card is only reachable after scrolling past all of that. Same root cause as the existing rule in `AGENTS.md` for a different dialog: "place a modal's primary content before a long scrolling picker... users should see how to complete the dialog without first discovering an off-screen footer" — just not yet applied here.

Both are real, both are worth fixing, and I've added them to the PR 3 prompt below. One thing worth deciding explicitly: `DecisionQualityTrendDialog` belongs to #719, not #28 — fixing its table is a reasonable thing to bundle into PR 3 since it shares the exact same "filter chrome eats the mobile viewport" pattern as `MistakeQueueDialog`, but it's touching a dialog this issue doesn't own. Flagging rather than deciding unilaterally.

**Comment by markafitzgerald1** [comment]
Claude Code (Opus 5, High effort), on Mark's behalf: both fast-follows are now filed, so the "not tracked anywhere" gap flagged earlier on this issue is closed.

- #765 — Classify why a discard was wrong, not just how much. This is the one the last acceptance criterion above deferred; the derivation from `allScoredKeepDiscardsByExpectedNetScoreDescending` is carried over verbatim.
- #766 — Let real-life hands entered by hand feed the mistake queue. New, from Mark's dogfooding: he enters hands he played with physical cards, and those are currently excluded from the queue by the `!isPractice` filter I recommended during #754's review. That recommendation was right for seeded and deep-linked hands and wrong for manual entry, because `EnterCardsDialog` submits six cards with no discards and the analysis only renders after the user picks two — so a manually entered hand is mechanically a blind first-instinct choice. #766 has the details and the correction.

Both are sequenced on the board after #763, which replaces the full-Docker pre-commit hook with a fast local check set. That ordering is deliberate: this session alone paid the duplicate-validation cost #763 describes six times over, once per review-fix commit, each running the full gate locally and then again in CI for the pushed SHA. Landing #763 first makes every subsequent PR — including both of these — measurably cheaper. Neither fast-follow blocks PR 3 of this issue.

**Comment by markafitzgerald1** [comment]
Claude Code (Sonnet 5, High effort), on Mark's behalf: plan for PR 3 of 3 — the practice drill — plus the two mobile layout fixes found during dogfooding. Read `skills/working-an-issue/SKILL.md` and the plan/status/mobile-bug comments above first; nothing below re-derives a decision already made there. Checked against the merged tree at e422abf.

## What PR 1 + 2 already built, that PR 3 only wires up

- `recordPracticeAttempt(attempt)` (`src/ui/discardTally.ts`) — writes only to `StoredTally.practice`, never `lifetime`; independently capped; safe against the record cap.
- `buildMistakeQueue(tally)` → `MistakeQueueItem[]` carrying `cards`, `cribRole`, `handKey`, `priority`, `pWrong`, `lossIfWrong`, `previousDiscard`, `consecutiveSuccesses`, `isMastered`.
- `MistakeQueueDialog`, rendered inside `DiscardTallyView`, opened from the "Mistake queue" button.
- The manual hand-load path (`Trainer.applyManualHand` → `reportHandReplaced(cards, "manual", role)`), which flags the loaded hand `isPractice: true` for free, so a drilled hand's re-score never touches the headline averages. `recordDiscardDecision` is also idempotent by `handKey`, and a drilled hand's key is by definition already present, so it writes nothing regardless.

So PR 3 adds no storage and no new math. It is: a start-drill action, an analysis gate, a verdict, one `recordPracticeAttempt` call, and a priority-weighted sampler.

## Drill wiring

New hook `src/ui-react/usePracticeDrill.ts`, extracted from `Trainer` following the `useAnalyticsConsent` / `useEnterCardsDialog` pattern already in that file — `Trainer.tsx` is 404 lines against the frozen 520-line cap and the state machine will not fit inline.

State: `activeItem: MistakeQueueItem | null`, `phase: "choosing" | "revealed"`, `verdict | null`, and a ref that guards a single `recordPracticeAttempt` per attempt (`onAnalysisRendered` re-fires on re-sorts).

- **Start** (`startDrill(item)` / `startAutoDrill()`): loads the hand through one seam, `loadDrillHand(item)`, which calls `applyManualHand({ cribRole: item.cribRole, dealtCards: toDealtCards(item.cards, []) })` — all six kept, the user picks the two. `startAutoDrill` first draws an item via the new sampler. This single seam is where #767's suit-permutation step later attaches; a comment marks it.
- **Gate**: `Trainer` renders `<ScoredPossibleKeepDiscards>` only when `discardIsComplete(dealtCards) && (!drill.isActive || drill.phase === "revealed")`. While `phase === "choosing"` with a complete discard, that slot renders nothing — a state both `.dynamic-ui` layouts already handle, identical to "no discard selected". No new positional child in `.dynamic-ui`; the revised-plan comment's point 7 is respected.
- **Commit**: a "Check discard" button — label checked against every existing control name for `getByRole` substring collisions — appears inside `InteractiveHand`, which already owns the role label and control row. Clicking it sets `phase = "revealed"`, mounting the analysis. `Trainer` wraps the existing `onAnalysisRendered` to also feed `drill.observeAnalysis(analysis)`.
- **Verdict**: `observeAnalysis` fires once (ref-guarded) when `phase === "revealed"` and `analysis.quality !== null`: it calls `recordPracticeAttempt({ at: Date.now(), handKey: toHandKey(dealtCards, cribRole), isOptimal, expectedPointsLoss })` and builds the verdict. The verdict banner, also inside `InteractiveHand`, says what changed (per the revised-plan comment's closing argument): the previous discard and its `lossIfWrong`, the discard chosen now and its `expectedPointsLoss`, and the delta; then whether this was optimal and how the mastery streak moved. Actions: **Next hand** (→ `startAutoDrill`) and **Exit drill**.
- **Exits**: Exit drill drops drill state, leaves the hand on the board, records nothing beyond what commit already wrote, and un-gates the analysis so it renders normally (honest degradation — a reload would reveal the answer anyway). Pressing Deal mid-drill is already correct by construction (the skip guard requires `practiceByHand.get(key) === false`, and a drill hand is flagged practice) — pinned by a test. Drill state stays out of the URL: `urlAnalysisState` is a public compatibility surface and a shared link must not start a stranger's drill.

## Start-drill entry points (`MistakeQueueDialog`)

- Per-item **"Drill this hand"** button on each mistake card.
- A top-level **"Start drill"** button that auto-draws by priority.
- Both call one `onStartDrill` prop; `DiscardTallyView` closes the dialog, then calls the `Trainer` callback.

## Priority-weighted sampler

New pure function in `src/ui/mistakeQueue.ts`: `sampleMistakeQueueByPriority(items, random /* [0,1) */): MistakeQueueItem | null` — filters to non-mastered, samples proportional to `priority`, falls back to uniform when every active item has `priority === 0`, returns `null` on an empty active set. `Trainer` supplies `random` from its existing `generateRandomNumber` prop. Same `priority` the list already sorts by — one function, two exposures, as the issue asks. No new constant.

## Mobile fix 1 — `MistakeQueueDialog` (this issue's dialog), at 375x812

Three summary cards + four filter groups + a description fill the viewport before the first mistake card. Apply the `AGENTS.md` rule ("place a modal's primary and secondary actions before a long scrolling picker and keep the action row sticky"): add a sticky action row (**Start drill** + **Close**) immediately after the title, ahead of the summary/filters/list, so the way to complete the dialog is visible on open. In portrait, drop `position: sticky` from `.controls` so two sticky rows do not stack.

## Mobile fix 2 — `DecisionQualityTrendDialog` (#719's dialog, homeless otherwise)

`.table-wrapper` has `overflow-y` but no `overflow-x`, so the Period/Batch table's Optimal and Skipped columns are cut off at phone width. CSS only: add `overflow-x: auto`, a `min-width` on `.table` sized to its five columns, and tighter `th`/`td` padding in the `aspect-ratio < 6/5` block. The PR body will flag this as touching a dialog #28 does not own, per the status comment.

## Out of scope, not precluded

- **#766** (any-provenance hands): the drill operates only on `MistakeQueueItem`, so it is already provenance-agnostic; no optimality assumptions added.
- **#767** (suit-permuted variant): the single `loadDrillHand` seam is the one place a permutation step will attach.
- **#765** (classify why a discard was wrong): untouched.

## Tests

- Unit: `sampleMistakeQueueByPriority` (distribution, empty, all-mastered, all-zero-priority); `usePracticeDrill` state machine and verdict construction.
- Component: `MistakeQueueDialog` start-drill buttons + sticky action row; `Trainer` drill flow (analysis hidden until commit, verdict after, `recordPracticeAttempt` called once, `lifetime` unmoved); `InteractiveHand` drill UI.
- Stories for the new `InteractiveHand` drill states and the `MistakeQueueDialog` action row.
- e2e (`tests-e2e/practiceDrill.spec.ts`, seeded tally in the `decisionQualityTrend.spec.ts` style): start a drill from the queue; analysis hidden until "Check discard"; verdict shows previous vs new discard; `.dynamic-ui` child count unchanged in drill mode; consent Accept still reachable at 844x390; a completed drill leaves the "Best choice" lifetime figure unmoved. Each guard negative-checked against the unfixed code.
- e2e for both mobile fixes at 375x812 / phone-landscape: MistakeQueueDialog action row visible without scrolling; trend table's rightmost column reachable.

## PR shape

Branch `feature/28-practice-drill`, pushed after one commit with the PR opened as a draft immediately, then the rest pushed. PR body carries a human review guide and a manual testing plan (a real phone at 375x812 for both mobile fixes; a full play → mistake → drill → re-drill loop in the preview). Copilot review requested; Codex loop run to a clean round before human review. The last line of the body closes #28 — this is the last of the three PRs.

**Comment by markafitzgerald1** [comment]
Claude Code (Sonnet 5, High effort), on Mark's behalf. For the record: after the PR #768 Codex loop went clean (round 9), dogfooding turned up two drill behaviours worth fixing, and both landed on the same branch.

**1. "Exit drill" now deals a fresh authentic hand.** It used to drop back to normal play with the drilled study hand still on the board and no live analysis wiring, so leaving practice left you staring at a hand you'd just dissected. `onExit` now splits into a bare `clearDrill` (state reset only — still what the Back/popstate path uses, since a history restore brings its own hand) and a wrapper that also asks `Trainer` for a fresh deal. Leaving practice always lands you in a normal playable position now.

**2. The sampler never serves the same hand twice in a row.** `sampleMistakeQueueByPriority` gained an optional `excludeHandKey`; `onNextHand` passes the hand it is leaving, so a "Draw another" run walks the open mistakes instead of possibly repeating one. Skipped when the drilled hand is the only active one left; an exhausted queue still exits to a fresh deal. Combined with the existing top-level "Start drill", this is the "just drill my past mistakes, interleaved, no thinking" loop from the third bullet of the dogfooding notes — "Start drill" then repeated "Draw another" is now hands-off, with no new UI needed.

Commits: `1ab3e13` (the two changes), `12f34ab` (AGENTS.md pointer fix from Codex round 10), `d18f14d` (test-name fix from the Copilot Lite review). CI `build-and-test` green on all; Codex round 11 clean.
